### ISSUE TITLE
Standardize Tombstoned bundles across schema versions.

### ISSUE THREAD

**Issue by Bento007** [HumanCellAtlas/data-store#1036]
**Story**: as a user i would like subscribe to tombstoned documents.

**Requirements:**
- [ ] A public json schema for tombstoned bunde.
- [ ] A query to search for tombstoned bundles.
 
**Definition of Done:**
Query the index and retrieve all tombstoned documents.

**Comment by ttung** [comment]
I would add that the schema for tombstones should be _versioned_ so that we have some notion of how to interpret tombstones should the version change.  Specifically, `_create_tombstone_data` should write some kind of a version, and the specifications should record this.

cc: @mweiden

**Comment by Bento007** [comment]
Is this something we should ask the Meta data team to design?

**Comment by hannes-ucsc** [comment]
@malloryfreeberg, do you think the tombstone schema should be maintained by the metadata group? A tombstone manifests itself in two places: (1) in the DSS storage bucket as a separate object whose name is derived from that of the bundle version to be deleted and (2) in the DSS index where that object's content (the JSON payload) takes the place of the deleted bundle version's content. The JSON payload could be used to convey the particulars of the deletion.

**Comment by kislyuk** [comment]
I think it would be a mistake to have tombstones presented as in-band documents, and by extension it is not appropriate to declare them as part of the schema. If we want to distinguish between "not found" and "has been deleted", we could use a custom HTTP error code like `410 Gone`.

**Comment by kislyuk** [comment]
In the context of events this would mean defining types of events (e.g. created, updated, deleted) out of band with respect to document metadata.

Also, as a general comment, "tombstone" is project-specific jargon that I think will be hard for users to understand, and we should avoid using it in user-facing documentation.

**Comment by hannes-ucsc** [comment]
There is value in giving an explanation of why an object that previously existed is now missing. That's why I like the idea of returning 410 in `GET /bundles/<uuid>` (the response body could be the tombstone JSON). `POST /search` is a different animal. I think users should be allowed to formulate a query that returns all deleted bundles. I don't really see any downsides to that. We don't need to be dogmatic about the tombstone shape, whether it is governed by a public schema or by a convention, as long as that choice is documented somewhere.  

BTW, the fact that the tombstone JSON replaces the tombstoned document in the index is not new. It was introduced last December (#684).

**Comment by malloryfreeberg** [comment]
Sorry for responding late to this (really busy with the preview release). I'm a little confused about what differentiates a "tombstone" schema from the collection of schemas that exist. We don't plan to delete any schemas, just upgrade them to newer versions as required. If a bundle is to be deleted, the bundle/JSON schemas that it uses will still exist (indicated by their versions).

**Comment by hannes-ucsc** [comment]
On storage a tombstone is an S3 object whose name is `<uuid>.<version>.dead` or `<uuid>.dead`. The former is supposed to make the bundle `<uuid>.<version>` inaccessible without deleting it. The latter is supposed to make all bundles starting in `<uuid>` inaccessible without deleting them. 

The content of that tombstone object is a JSON document. In the ES index, that document replaces the content of the bundle or bundles. 

It is my opinion that if tombstones contain JSON, the format of that JSON should be standardized using a JSON schema.

**Comment by ttung** [comment]
I would like to emphasize that tombstones are an internal blue box implementation detail.  Its schema should *definitely* not be maintained by the metadata group, which is concerned with _biological_ metadata.

**Comment by hannes-ucsc** [comment]
> which is concerned with biological metadata.

Not entirely true:

https://github.com/HumanCellAtlas/metadata-schema/blob/master/json_schema/bundle/submission.json

**Comment by ttung** [comment]
That's an external interface that people will need to understand, as I understand.  The contents of the tombstone file is not, as long as the external interface is defined.

**Comment by hannes-ucsc** [comment]
IIRC, the reason for the deletion of a bundle was supposed to be explained to external users. I could be wrong.

**Comment by ttung** [comment]
Right, I would assume that would be returned in GET /bundles or whatever.  It's fine to have a schema for that, which is already more or less specified in `dss-api.yml`.  The schema for the tombstone is much like the schema for `/bundles/<bundlefqid>` or `/files/<filefqid>` (the blobs themselves, not the endpoints).  Should be tracked, yes, not part of the metadata working group's concern.

**Comment by malloryfreeberg** [comment]
Thanks for the explanation @hannes-ucsc! I think I understand now. So the tombstone object would live with the *tombstoned* object and would be what is returned to users if they try to search/query for the original object (which is now marked as dead). The point of this issue, then, is to define what that tombstone document will look like i.e. what its schema should be.

Question: Do tombstone objects need to reflect or indicate the contents of the original object? In other words, do we need to have a biomaterial tombstone object/schema, a file tombstone object/schema, etc for each metadata entity?

I think I agree with @ttung that the metadata-schema repo should not contain these schemas (my reasoning is that these schemas will never be created by users, only by blue box), and I also agree with @hannes-ucsc that there is precedence for external users to interact with this object as it will be what is returned to them when they search for the dead object. It would be ideal if the tombstone schemas were aligned with the actual metadata schemas, and to this end I'm happy to be involved in the discussion and development of them. I'm leaning towards the metadata team not being in charge of maintaining them, though.

**Comment by hannes-ucsc** [comment]
> these schemas will never be created by users, only by blue box

An admin user (or any user with delete permission) should provide the tombstone document at deletion time. They are users. Maybe not external ones, but still users.

> Should be tracked, yes, not part of the metadata working group's concern.

Please elaborate why not. The original claim was that metadata-schema repo is only concerned with biological information, which is not true. The reply was that users are not concerned with this data, which sounds like it is also not true, if I remember the use cases correctly.

Tombstones are data submitted by users and consumed by users. They contain information about non-technical aspects of a deletion, like the reason behind it. To me its obvious that this falls into the information realm for which the metadata-schema repo hosts schemas.

**Comment by ttung** [comment]
I would strongly advise that the _interface_ be what is controlled and defined, and the actual contents be an implementation decision.

**Comment by hannes-ucsc** [comment]
Duly noted. But without any motivation or reasoning behind that advice, I can't give it much weight. Sometimes it is preferable to leave the content up to the implementation detail, sometimes it isn't.

**Comment by malloryfreeberg** [comment]
> An admin user (or any user with delete permission) should provide the tombstone document at deletion time. They are users. Maybe not external ones, but still users.

I guess I'm unsure how tombstoning works. I was picturing that a user (in this context, someone who has submitted data to the HCA) would realize they needed to delete some data (for whatever reason). They then put a request in through the UI (or email the helpdesk) describing the reason for the deletion, and at that point there is some mechanism in the data store that identifies the files to be tombstoned, and automatically generates the new tombstone object that replaces the files to be deleted. In this scenario, the user doesn't need to know that a tombstone document is being generated, and thus doesn't need to know/understand the schema.

Is what I've described accurate? Is there another scenario in which a user has to actually create the tombstone object directly rather than it being automatically generated? I'm not trying to be argumentative, I just want to understand the process :)

**Comment by hannes-ucsc** [comment]
That accurately describes the current implementation with one caveat. The Tombstone object's content replaces the deleted bundle object's document in the index. If you search for a deleted bundle's UUID, you get back the tombstone object's content. You can even search for tombstones in the index:

```
{
    "query": {
        "bool": {
            "must": [
                {
                    "term": {
                        "admin_deleted": true
                    }
                }
            ]
        }
    }
}
```

This gives
```
{
  "took": 2,
  "timed_out": false,
  "_shards": {
    "total": 30,
    "successful": 30,
    "failed": 0
  },
  "hits": {
    "total": 2646,
    "max_score": 0.0012414651,
    "hits": [
      {
        "_index": "dss-prod-aws-docs-v.biomaterial.5.file.1.links.1.process.5.project.5.protocol.5",
        "_type": "doc",
        "_id": "17a3d288-01a0-464a-9599-7375fda3353d.2018-03-28T150956.976400Z",
        "_score": 0.0012414651,
        "_source": {
          "email": "travis-test@human-cell-atlas-travis-test.iam.gserviceaccount.com",
          "reason": "",
          "admin_deleted": true,
          "version": "2018-03-28T150956.976400Z",
          "uuid": "17a3d288-01a0-464a-9599-7375fda3353d"
        }
      },
…
```

This is where the tombstone document becomes evident to external users. Neither `reason`, `email` or `admin_deleted` are standardized or documented anywhere. Reason might need to become a structured field in the future.

**Comment by hannes-ucsc** [comment]
It just occurred to me that the top-level, skeletal shape of bundle documents in data-store's ES index is not specified anywhere either. Properties like `files.foobar_json` or `manifest.version` are not subject to any schema. I think tombstone properties such like `email`, `admin_deleted` and `reason` probably fall into the same category of properties. While they are visible to external users and probably should be controlled by a schema, I've come around to thinking that that schema should NOT be managed by the metadata WG.
